File: app/javascript/javascripts/modules/dropdown/toggle_param_item.js
// Copyright (c) 2022, Die Mitte Schweiz. This file is part of
// hitobito and licensed under the Affero General Public License version 3
// or later. See the COPYING file at the top-level directory or at
// https://github.com/hitobito/hitobito.

// check app/helpers/dropdown/toggle_param_item.rb

const ToggleParamItem = {

  get toggleParamLinks() {
    return document
      .querySelectorAll('.dropdown-menu a.toggle-param');
  },
  
  toggleChecked(event) {
    // this prevents closing of drop down
    // when clicking on link
    event.stopPropagation();
    event.preventDefault();
    const link = event.target;
    const checkbox = link.querySelector('input[type="checkbox"]');
    checkbox.checked = !checkbox.checked;
    this.updateLinkParams(checkbox);
  },

  registerToggleParamLinks() {
    this.toggleParamLinks.forEach((toggleLink) => {
      toggleLink.addEventListener('click', (event) => {
        this.toggleChecked(event);
      })
    })
  },

  get toggleParamCheckboxes() {
    return document
      .querySelectorAll('.toggle-param input[type="checkbox"]');
  },

  linkItems(checkbox) {
    const parent = checkbox.closest('.dropdown-menu');
    return parent.querySelectorAll('a:not(.toggle-param)');
  },

  updateLinkParams(checkbox) {
    const paramName = checkbox.dataset.toggleParamId;
    const checked = checkbox.checked;
    this.linkItems(checkbox).forEach((link) => {
      let href = link.getAttribute('href');
      href = href.replace(`${paramName}=${!checked}`, `${paramName}=${checked}`)
      link.setAttribute('href', href);
    })
  },

  register() {
    this.registerToggleParamLinks();
  }
}

document.addEventListener("DOMContentLoaded",
    () => ToggleParamItem.register());

File: app/helpers/dropdown/toggle_param_item.rb
# frozen_string_literal: true

#  Copyright (c) 2022, Die Mitte Schweiz. This file is part of
#  hitobito and licensed under the Affero General Public License version 3
#  or later. See the COPYING file at the top-level directory or at
#  https://github.com/hitobito/hitobito.

module Dropdown
  class ToggleParamItem

    def initialize(template, id, label, checked: false)
      @template = template
      @id = id
      @label = label
      @checked = checked
    end

    def render(template)
      template.content_tag(:li) do
        template.link_to('#', class: "toggle-param") do
          render_checkbox(template)
        end
      end
    end

    def render_checkbox(template)
      template.content_tag(:div, class: 'checkbox') do
        template.content_tag(:label, for: @id) do
          template.safe_join([
            template.check_box_tag(@id, '1', @checked, 'data-toggle-param-id': @id),
            @label
          ].compact)
        end
      end
    end
  end
end
